Fixing RPC Shutdown and Thread Joining

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36239 Fixing RPC Shutdown and Thread Joining**

ProcessGroupAgent and ThriftAgent threads were joined at shutdown, but RpcAgent threads were joined by the destructor. This PR joins all threads at shutdown by using a pattern similar to `start` in RPC.

The derived classes implement a `shutdownImpl` class that cleans up backend-specific state. RpcAgent implements `shutdown` which cleans up generic state and calls the underlying `shutdownImpl`. The atomic running is now set and unset by RpcAgent so backends do not need to mutate it.

Fixes: https://github.com/pytorch/pytorch/issues/36242

Differential Revision: [D20902666](https://our.internmc.facebook.com/intern/diff/D20902666/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20902666/)!

osalpekar committed:
Fixing RPC Shutdown and Thread Joining

ProcessGroupAgent and ThriftAgent threads were joined at shutdown, but RpcAgent threads were joined by the destructor. This PR joins all threads at shutdown by using a pattern similar to `start` in RPC.

The derived classes implement a `shutdownImpl` class that cleans up backend-specific state. RpcAgent implements `shutdown` which cleans up generic state and calls the underlying `shutdownImpl`. The atomic running is now set and unset by RpcAgent so backends do not need to mutate it.

Differential Revision: [D20902666](https://our.internmc.facebook.com/intern/diff/D20902666/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20902666/)!

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -32,14 +34,13 @@ void RpcAgent::start() {
 
-void RpcAgent::cleanup() {
+void RpcAgent::shutdown() {
   rpcAgentRunning_.store(false);
```
I think there's actually a race condition here - you might need to temporarily grab rpcRetryMutex_ when setting for this var for the exit to work properly.

(I just noticed a similar issue in a ParallelNative.cpp PR yesterday)

It's non-intuitive. but consider the following:
```
  - retryThread:       if (!rpcAgentRunning_.load())    // still true, so we do nothing
  - rpcAgent:     rpcAgentRunning_.store(false);	  // no it's false
                             rpcRetryMapCV_.notify_one();	  // no waiters in the cv, so it does nothing
   - retryThread:       rpcRetryMapCV_.wait_until(lock, earliestTimeout);   // shoot, we basically sleep forever if it's empty    

Where the mutex adds value is  in  protecting against mutation in the critical section between the waiter noticing one value and going to sleep.

This seems to be somewhat related to the purpose of your change here, do you want to fix along with this?

osalpekar replied:
I see, so in the above case, `shutdown` will only grab rpcRetryMutex_ when the retry thread unlocks the lock upon entering the `wait_until` - so the var can't be set to False between checking the value and the retry thread sleeping. Makes sense, I'll fix this.

osalpekar committed:
Update on "Fixing RPC Shutdown and Thread Joining"


ProcessGroupAgent and ThriftAgent threads were joined at shutdown, but RpcAgent threads were joined by the destructor. This PR joins all threads at shutdown by using a pattern similar to `start` in RPC.

The derived classes implement a `shutdownImpl` class that cleans up backend-specific state. RpcAgent implements `shutdown` which cleans up generic state and calls the underlying `shutdownImpl`. The atomic running is now set and unset by RpcAgent so backends do not need to mutate it.

Fixes: https://github.com/pytorch/pytorch/issues/36242

Differential Revision: [D20902666](https://our.internmc.facebook.com/intern/diff/D20902666/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20902666/)!

[ghstack-poisoned]

osalpekar committed:
Update on "Fixing RPC Shutdown and Thread Joining"


ProcessGroupAgent and ThriftAgent threads were joined at shutdown, but RpcAgent threads were joined by the destructor. This PR joins all threads at shutdown by using a pattern similar to `start` in RPC.

The derived classes implement a `shutdownImpl` class that cleans up backend-specific state. RpcAgent implements `shutdown` which cleans up generic state and calls the underlying `shutdownImpl`. The atomic running is now set and unset by RpcAgent so backends do not need to mutate it.

Fixes: https://github.com/pytorch/pytorch/issues/36242

Differential Revision: [D20902666](https://our.internmc.facebook.com/intern/diff/D20902666/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20902666/)!

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -740,13 +741,13 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
       // the predicate missed the notification before it started waiting
       // on the cond var.
-      if (!rpcAgentRunning_.load()) {
+      if (!timeoutThreadEnabled_.load()) {
```
maybe you also update rpcAgentRunning_ in line 730 as well to this var?

(technically it would work as-is, but if the underlying assumption ever changes, we end up with a busy spin here)

osalpekar committed:
Update on "Fixing RPC Shutdown and Thread Joining"


ProcessGroupAgent and ThriftAgent threads were joined at shutdown, but RpcAgent threads were joined by the destructor. This PR joins all threads at shutdown by using a pattern similar to `start` in RPC.

The derived classes implement a `shutdownImpl` class that cleans up backend-specific state. RpcAgent implements `shutdown` which cleans up generic state and calls the underlying `shutdownImpl`. The atomic running is now set and unset by RpcAgent so backends do not need to mutate it.

Fixes: https://github.com/pytorch/pytorch/issues/36242

Differential Revision: [D20902666](https://our.internmc.facebook.com/intern/diff/D20902666/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20902666/)!

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -210,7 +210,12 @@ class TORCH_API RpcAgent {
   // Stop accepting requests and shutdown the RPC framework as soon as possible
   // by terminating all RPC threads.
-  virtual void shutdown() = 0;
+  virtual void shutdown();
```
Sorry, just noticed this - does it make sense to make it non-virtual?  (and if that's the case, to comment on it being on purpose)
I thought the idea is that subclasses would override shutdownImpl(), not this method?

osalpekar replied:
It should be non-virtual, I'll update the PR.

osalpekar committed:
Update on "Fixing RPC Shutdown and Thread Joining"


ProcessGroupAgent and ThriftAgent threads were joined at shutdown, but RpcAgent threads were joined by the destructor. This PR joins all threads at shutdown by using a pattern similar to `start` in RPC.

The derived classes implement a `shutdownImpl` class that cleans up backend-specific state. RpcAgent implements `shutdown` which cleans up generic state and calls the underlying `shutdownImpl`. The atomic running is now set and unset by RpcAgent so backends do not need to mutate it.

Fixes: https://github.com/pytorch/pytorch/issues/36242

Differential Revision: [D20902666](https://our.internmc.facebook.com/intern/diff/D20902666/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20902666/)!

[ghstack-poisoned]
